test/Generics/where_clause_contextually_generic_decls.swift

```diff
@@ -0,0 +1,120 @@
+    where Self: Sequence {
+      return arg
+  }
+
+  func protoFunc() where Assoc == Bssoc {}
+  // expected-note@-1 {{where 'Self.Assoc' = 'T1.Assoc', 'Self.Bssoc' = 'T1.Bssoc'}}
+
+  // expected-note@+1 {{where 'Self.Assoc' = 'T1.Assoc'}}
+  subscript() -> Assoc? where Assoc: Whereable {
+      return nil
+  }
+}
+
+class HelperClass {}
+class Class<T> {
+  class func classFunc1() where T: Whereable, T.Assoc == Int {}
+  // expected-note@-1 {{where 'T' = 'T}}
+
+  func func1() where T == Bool {}
+  // expected-note@-1 {{candidate requires that the types 'T' and 'Bool' be equivalent}}
+  func func1() where T == Int {}
+  // FIXME: The rhs type at the end of the error message is not persistent across compilations.
+  // expected-note@-2 {{candidate requires that the types 'T' and 'Int' be equivalent (requirement specified as 'T' == }}
+
+  func func2() where T == Int {} // expected-note {{where 'T' = 'T'}}
+
+  // expected-note@+1 {{where 'T' = 'T'}}
+  subscript() -> T.Element? where T: Sequence {
+    return nil
+  }
+
+  typealias NestedAlias1 = Bool where T == Int
+  // expected-note@-1 {{'NestedAlias1' previously declared here}}
+  typealias NestedAlias1 = Bool where T == Bool
+  // expected-error@-1 {{invalid redeclaration of 'NestedAlias1}}
+  typealias NestedAlias2 = T.Assoc where T: Whereable
+
+  enum NestedEnum1<U> {}
+  class NestedClass1 where T: Whereable {}
+}
+
+extension Class.NestedEnum1 {
+  func foo(arg: U) where U: Whereable, U.Assoc == T { }
+}
+
+extension Class {
+  class func classFunc2() where T: HelperClass {} // expected-note {{where 'T' = 'T'}}
+
+  subscript(arg: T.Element) -> T.Element? where T == Array<Int> {
+    return nil
+  }
+}
+
+extension Class where T: Equatable {
+  func func4() where T: Comparable { } // expected-note {{where 'T' = 'T'}}
+
+  func badFunc1() where T == Class<Int> { }
+  // expected-error@-1 {{same-type constraint type 'Class<Int>' does not conform to required protocol 'Equatable'}}
+}
+
+extension Class where T == Bool {
+  func badFunc2() where T == Int { }
+  // expected-error@-1 {{generic parameter 'T' cannot be equal to both 'Int' and 'Bool'}}
+}
+
+func testMemberDeclarations<T, U: Comparable>(arg: Class<T>, arg1: Class<U>, t: T) {
+  Class<T>.classFunc1() // expected-error {{class method 'classFunc1()' requires that 'T' conform to 'Whereable'}}
+  Class<T>.classFunc2() // expected-error {{class method 'classFunc2()' requires that 'T' inherit from 'HelperClass'}}
+  arg.func1() // expected-error {{no exact matches in call to instance method 'func1()'}}
+  arg.func2() // expected-error {{instance method 'func2()' requires the types 'T' and 'Int' be equivalent}}
+  arg.func4() // expected-error {{instance method 'func4()' requires that 'T' conform to 'Comparable'}}
+  arg[] // expected-error {{subscript 'subscript()' requires that 'T' conform to 'Sequence'}}
+
+  // FIXME: None of these should be accessible on Class<T>.
```

Comment:
I recently fixed this for generic type decls -- it should be easy to adapt the logic to handle the new case

Reply:
See #29544